### ISSUE TITLE
Memory GPU leak with Gluon

### ISSUE THREAD

**Issue by edmBernard** [apache/incubator-mxnet#8747]
I try to train a simple vgg16 on large custom database.
with this training loop:
```python
for data, label in self.base.train:
  print(label, end=" ")
  # moving datas and labels in the net's context
  data = data.as_in_context(ctx)
  label = label.as_in_context(ctx)
  # computing forward pass in the autograd environement (for gradient computation)
  with autograd.record():
    output = net(data)
    loss = scel(output, label)
    # loss back-propagation
  loss.backward()
  # trainer update
  trainer.step(data.shape[0])
  # ndarray.waitall()
```

With only this the GPU memory slowly increase until GPU crash.

If I set `MXNET_ENGINE_TYPE=NaiveEngine` memory leak disappear and if I add `ndarray.waitall()` after each batch the memory leak disappear too.

Any clue how to manage this ?

**Comment by whr94621** [comment]
I think a ndarray.waitall() is necessary after each batch. Gluon default uses lazy evaluation, if you don't print the loss, use asnumpy() or use a waitall(), gluon actually adds more data to computation graph instead of doing computation and release the memory.

**Comment by edmBernard** [comment]
Thanks that's what I suppose but I thought there was a guard to avoid GPU saturation.

**Comment by jiarenyf** [comment]
@whr94621 Solve my problem,  thank you ...
